### DISCUSSION TITLE
Any way to ensure GPU as a device with NBEATSx example usage?

### DISCUSSION THREAD

**Discussion by rcslight** [Nixtla/neuralforecast#572] [Q&A]
Hello everyone!

I've been following the [NBEATSx example usage](https://github.com/Nixtla/neuralforecast/blob/main/nbs/models.nbeatsx.ipynb) but can't seem to enforce GPU as a device. I have scrambled the lib and docs and didn't find an evident way to do it, or if it's even possible.

I've tried adding GPU as accelerator to no avail:

```python
model2 = NBEATSx(h=12, input_size=24,
                loss=DistributionLoss(distribution='Normal', level=[80, 90]),
                scaler_type='robust',
                stat_exog_list=['airline1'],
                futr_exog_list=['trend'],
                max_steps=200,
                val_check_steps=10,
                early_stop_patience_steps=2,
                accelerator='gpu',)

model2.device
```
```
> Global seed set to 1
> device(type='cpu')
```

I can use it on the same environment with pytorch without any trouble.

```
import torch
torch.cuda.is_available()
```
```
> True
```

I've been having fun with the lib so far on my usecase, and hope to contribute with yalls in the future! Cheers!

**Comment by kdgutier** [comment]
Hey @rcslight,

Thanks for using NeuralForecast, by default our methods will check if you have a GPU available and execute with it.
If you are using Google Colab, make sure that your instance is a GPU here:
![Screenshot 2023-05-12 at 10 23 01 AM](https://github.com/Nixtla/neuralforecast/assets/19935241/74209457-5469-4c4f-b5b5-915b367a417b)

If you are using your own gpu, you have to make sure that you have a correct CUDA installation, and that your python environment detects your GPU.
https://stackoverflow.com/questions/48152674/how-do-i-check-if-pytorch-is-using-the-gpu
```
>>> import torch
>>> torch.cuda.is_available()
True
```

**Comment by rcslight** [comment]
hi, thanks for the insanely quick answer!

I'm using my own device, and it all checks out, however whenever I instantiate the model it still default to CPU it seems.

![image](https://github.com/Nixtla/neuralforecast/assets/90638822/1e41468f-d6f3-45d6-a348-242cf10af4bc)

![image](https://github.com/Nixtla/neuralforecast/assets/90638822/2480ff10-fee4-4688-aa31-ec7fb0915d4c)


It happens that when I test it with my own dataset, funnily enough, it shows up as `cuda` and also in GPU memory when monitoring with `nvidia-smi`:

![image](https://github.com/Nixtla/neuralforecast/assets/90638822/45facb59-f25d-4890-9510-972d9ae7ccaa)

After trying to check what cause these observations, I was able to nail down the behavior by pausing the training and checking for the device.

- When the model run all epochs:

![image](https://github.com/Nixtla/neuralforecast/assets/90638822/87a140da-582e-4c80-891b-ab34a58b8f39)

- When paused during epoch 0:

![image](https://github.com/Nixtla/neuralforecast/assets/90638822/67091441-aa01-4a34-bd41-2932bfd2520f)

maybe `nf.models[0].device` defaults to `'cpu'` when unused?

**Comment by kdgutier** [comment]
To save the GPU memory usage
the instantiated models will live in the CPU memory until called by the Pytorch Lightning trainer that moves the models to the GPU during the execution of `neuralforecast.fit` and `neuralforecast.predict`.

From your readings of the NVIDIA/SMI you are already using the GPU.

**Comment by rcslight** [comment]
Great, thanks for the assistance!
I'll close the discussion.
